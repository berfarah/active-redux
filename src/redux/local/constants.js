export const LOCAL_CLEAR = 'AR_LOCAL_CLEAR';
export const LOCAL_HYDRATE = 'AR_LOCAL_HYDRATE';
export const LOCAL_READ = 'AR_LOCAL_READ';
export const LOCAL_CREATE = 'AR_LOCAL_CREATE';
export const LOCAL_UPDATE = 'AR_LOCAL_UPDATE';
export const LOCAL_DELETE = 'AR_LOCAL_DELETE';
