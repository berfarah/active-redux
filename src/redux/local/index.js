export * from './actions';
export localReducer from './reducer';
