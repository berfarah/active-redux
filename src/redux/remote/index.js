export * from './actions';
export remoteReducer from './reducer';
