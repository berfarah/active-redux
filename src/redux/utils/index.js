import imm from 'object-path-immutable';

export apiClient from './api-client';

export const incrementProperty = (state, key) => (
  imm.set(state, key, state[key] + 1)
);

export const decrementProperty = (state, key) => (
  imm.set(state, key, state[key] - 1)
);

export const resourcesArray = (data) => (
  Array.isArray(data) ? data : [data]
);

export const mergeResources = (state, { data, included = [] }) => {
  const newState = imm(state);

  resourcesArray(data).concat(included).forEach((dataObj) => {
    // if we don't do this and the ID is a number, it'll create an array
    newState.set(dataObj.type, state[dataObj.type] || {});
    newState.set(`${dataObj.type}.${dataObj.id}`, dataObj);
  });

  return newState.value();
};

export const clearResources = (state, resources) => {
  switch (typeof resources) {
    case 'object': {
      if (Array.isArray(resources)) return resources.reduce(clearResources, state);
      const { type, id } = resources;
      return imm.del(state, ['resources', type, id]);
    }
    case 'string': return imm.set(state, ['resources', resources], {});
    case 'undefined': return imm.set(state, 'resources', {});
    default: return state;
  }
};

export const createAction = (type) => (payload) => ({ type, payload });

export const createReducer = (mappings, initialState) => (state = initialState, action) => (
  mappings[action.type] ? mappings[action.type](state, action) : state
);
