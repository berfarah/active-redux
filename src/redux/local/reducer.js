import imm from 'object-path-immutable';

import * as ActionTypes from './constants';
import {
  resourcesArray,
  mergeResources,
  clearResources,
  createReducer,
} from '../utils';

const updateResources = (state, resources) => (
  imm.set(state, 'resources', mergeResources(state.resources, resources))
);

const initialState = {
  resources: {},
};

export default createReducer({
  [ActionTypes.LOCAL_CLEAR]: (state, { payload: type }) => (
    type ? clearResources(state, type) : state
  ),
  [ActionTypes.LOCAL_HYDRATE]: (state, { payload: resources }) => (
    updateResources(state, resources)
  ),
  [ActionTypes.LOCAL_CREATE]: (state, { payload: resources }) => (
    updateResources(state, resources)
  ),
  [ActionTypes.LOCAL_READ]: (state, { payload: resources }) => (
    updateResources(state, resources)
  ),
  [ActionTypes.LOCAL_UPDATE]: (state, { payload: resources }) => (
    updateResources(state, resources)
  ),
  [ActionTypes.LOCAL_DELETE]: (state, { payload: resources }) => (
    clearResources(state, resourcesArray(resources.data))
  ),
}, initialState);
