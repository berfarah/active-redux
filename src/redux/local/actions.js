import * as ActionTypes from './constants';
import { createAction } from '../utils';

export const localClear = createAction(ActionTypes.LOCAL_CLEAR);
export const localHydrate = createAction(ActionTypes.LOCAL_HYDRATE);
export const localCreate = createAction(ActionTypes.LOCAL_CREATE);
export const localRead = createAction(ActionTypes.LOCAL_READ);
export const localUpdate = createAction(ActionTypes.LOCAL_UPDATE);
export const localDelete = createAction(ActionTypes.LOCAL_DELETE);
