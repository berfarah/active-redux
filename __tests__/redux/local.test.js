import * as subject from 'active-redux/redux/local';

describe('localReducer', () => {
  const person = {
    type: 'people',
    id: 5,
    attributes: {
      hair: 'black',
      age: 25,
    }
  };

  describe('initialState', () => {
    it('has an empty resources object', () => {
      expect(subject.localReducer(undefined, {})).toMatchSnapshot();
    });
  });

  describe('actions', () => {
    const reduce = subject.localReducer;

    ['localHydrate', 'localCreate', 'localRead'].forEach((actionName) => {
      describe(`${actionName}()`, () => {
        const action = subject[actionName]({ data: person });

        it('populates the state with resources', () => {
          expect(reduce(undefined, action)).toMatchSnapshot();
        });
      });
    });

    describe('localClear()', () => {
      const initialState = { resources: { [person.type]: { 1: 'foo', 2: 'bar' } } };
      const action = subject.localClear(person.type);

      it('clears the resource from state', () => {
        expect(reduce(initialState, action)).toMatchSnapshot();
      });
    });

    describe('localUpdate()', () => {
      const initialState = { resources: { [person.type]: { 1: 'foo', 2: 'bar' } } };
      const action = subject.localUpdate({ data: person });

      it('updates that resource', () => {
        expect(reduce(initialState, action)).toMatchSnapshot();
      });
    });

    describe('localDelete()', () => {
      const initialState = {
        resources: {
          [person.type]: {
            1: 'foo',
            2: 'bar',
            [person.id]: person
          }
        }
      };
      const action = subject.localDelete({ data: person });

      it('deletes that resource', () => {
        expect(reduce(initialState, action)).toMatchSnapshot();
      });
    });
  });
});
