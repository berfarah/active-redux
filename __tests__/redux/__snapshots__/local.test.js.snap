// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`localReducer actions localClear() clears the resource from state 1`] = `
Object {
  "resources": Object {
    "people": Object {},
  },
}
`;

exports[`localReducer actions localCreate() populates the state with resources 1`] = `
Object {
  "resources": Object {
    "people": Object {
      "5": Object {
        "attributes": Object {
          "age": 25,
          "hair": "black",
        },
        "id": 5,
        "type": "people",
      },
    },
  },
}
`;

exports[`localReducer actions localDelete() deletes that resource 1`] = `
Object {
  "resources": Object {
    "people": Object {
      "1": "foo",
      "2": "bar",
    },
  },
}
`;

exports[`localReducer actions localHydrate() populates the state with resources 1`] = `
Object {
  "resources": Object {
    "people": Object {
      "5": Object {
        "attributes": Object {
          "age": 25,
          "hair": "black",
        },
        "id": 5,
        "type": "people",
      },
    },
  },
}
`;

exports[`localReducer actions localRead() populates the state with resources 1`] = `
Object {
  "resources": Object {
    "people": Object {
      "5": Object {
        "attributes": Object {
          "age": 25,
          "hair": "black",
        },
        "id": 5,
        "type": "people",
      },
    },
  },
}
`;

exports[`localReducer actions localUpdate() updates that resource 1`] = `
Object {
  "resources": Object {
    "people": Object {
      "1": "foo",
      "2": "bar",
      "5": Object {
        "attributes": Object {
          "age": 25,
          "hair": "black",
        },
        "id": 5,
        "type": "people",
      },
    },
  },
}
`;

exports[`localReducer initialState has an empty resources object 1`] = `
Object {
  "resources": Object {},
}
`;
